mysqld_exporter.go

```diff
@@ -0,0 +1,179 @@
+package main
+
+import (
+	"flag"
+	"log"
+	"net/http"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"database/sql"
+	_ "github.com/go-sql-driver/mysql"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	namespace = "mysql"
+)
+
+var (
+	listenAddress = flag.String("web.listen-address", ":9104", "Address to listen on for web interface and telemetry.")
+	metricPath    = flag.String("web.telemetry-path", "/metrics", "Path under which to expose metrics.")
+)
+
+type Exporter struct {
+	dsn                        string
+	mutex                      sync.RWMutex
+	duration                   prometheus.Gauge
+	totalScrapes, errorScrapes prometheus.Counter
+	metrics                    map[string]prometheus.Gauge
+}
+
+// return new empty exporter
+func NewMySQLExporter(dsn string) *Exporter {
+	return &Exporter{
+		dsn: dsn,
+		duration: prometheus.NewGauge(prometheus.GaugeOpts{
+			Namespace: namespace,
+			Name:      "exporter_last_scrape_duration_nanoseconds",
+			Help:      "Was the last scrape  duration.",
+		}),
+		totalScrapes: prometheus.NewCounter(prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "exporter_scrapes_total",
+			Help:      "Current total mysqld scrapes.",
+		}),
+		errorScrapes: prometheus.NewCounter(prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "exporter_scrape_errors_total",
+			Help:      "Error mysqld scrapes.",
+		}),
+		metrics: map[string]prometheus.Gauge{},
+	}
+}
+
+func (e *Exporter) Describe(ch chan<- *prometheus.Desc) {
+	for _, m := range e.metrics {
+		m.Describe(ch)
+	}
+
+	ch <- e.duration.Desc()
+	ch <- e.totalScrapes.Desc()
+	ch <- e.errorScrapes.Desc()
+}
+
+func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
+	scrapes := make(chan []string)
+
+	go e.scrape(scrapes)
+
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+	e.setMetrics(scrapes)
+	ch <- e.duration
+	ch <- e.totalScrapes
+	ch <- e.errorScrapes
+	e.collectMetrics(ch)
+}
+
+func (e *Exporter) scrape(scrapes chan<- []string) {
+	defer close(scrapes)
+
+	now := time.Now().UnixNano()
+
+	e.totalScrapes.Inc()
+
+	db, err := sql.Open("mysql", e.dsn)
+	if err != nil {
+		log.Printf("error open connection to db")
+		e.errorScrapes.Inc()
+		e.duration.Set(float64(time.Now().UnixNano() - now))
+		return
+	}
+	defer db.Close()
+
+	rows, err := db.Query("SHOW STATUS")
+	if err != nil {
+		log.Printf("error running query on db")
```

Comment:
I'd also log the error message:

    log.Println("error running status query on database:", err)